### PR TITLE
Save flash on generated struct encoding logic

### PR DESCRIPTION
**Pull request by tcarmelveilleux**
Problem:

- Structural repetition in the generated code caused needless jumps to be inlined in very large numbers in `Encode` methods, in ways that could be factored-out.

This PR:

- Changes the generated code to use a proxy encoder class that reduces expanded inlined code, reduces 64 bit arguments usage and maintains correct early return behavior, while generating significantly smaller method bodies.

Testing done:

- Unit tests pass
- Integration tests pass

### CODE REVIEW

File: src/app/data-model/WrappedStructEncoder.h [hunk 1]
```diff
@@ -0,0 +1,71 @@
+/*
+ *
```
Comment (bzbarsky-apple) [inline src/app/data-model/WrappedStructEncoder.h:2]:
```suggestion
```

File: src/app/data-model/WrappedStructEncoder.h [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <app/data-model/Encode.h>
+#include <app/data-model/FabricScoped.h>
+#include <app/data-model/List.h>
+#include <app/data-model/Nullable.h>
```
Comment (bzbarsky-apple) [inline src/app/data-model/WrappedStructEncoder.h:24]: None of these are used here anymore, right?

File: src/app/data-model/WrappedStructEncoder.h [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+
+#include <lib/core/CHIPError.h>
+#include <lib/core/DataModelTypes.h>
+#include <lib/core/Optional.h>
```
Comment (bzbarsky-apple) [inline src/app/data-model/WrappedStructEncoder.h:28]: Optional is not used here.

File: src/app/data-model/WrappedStructEncoder.h [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+#include <lib/core/TLV.h>
+#include <protocols/interaction_model/Constants.h>
```
Comment (bzbarsky-apple) [inline src/app/data-model/WrappedStructEncoder.h:30]: I don't think we are using anything from Constants.h here either.

**Review by bzbarsky-apple — APPROVED** [review]

**Review by yufengwangca — APPROVED** [review]

File: src/app/data-model/WrappedStructEncoder.h [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    {
+        VerifyOrReturn(mLastError == CHIP_NO_ERROR);
+        mLastError = DataModel::Encode(mWriter, TLV::ContextTag(contextTag), std::forward<Args>(args)...);
+    }
+
+    CHIP_ERROR Finalize()
+    {
+        if (mLastError == CHIP_NO_ERROR)
+        {
+            mWriter.EndContainer(mOuter);
```
Comment (Damian-Nordic) [inline src/app/data-model/WrappedStructEncoder.h:51]:
```suggestion
            mLastError = mWriter.EndContainer(mOuter);
```

**Review by Damian-Nordic — APPROVED** [review]
Nice result! :)

**Pull request merged**